src/renderer/dx/DxFontRenderData.h

```diff
@@ -74,10 +77,15 @@ namespace Microsoft::Console::Render
 
         [[nodiscard]] static HRESULT STDMETHODCALLTYPE s_CalculateBoxEffect(IDWriteTextFormat* format, size_t widthPixels, IDWriteFontFace1* face, float fontScale, IBoxDrawingEffect** effect) noexcept;
 
+        bool DidUserSetFeatures() const noexcept;
+        void SetFeatures(std::unordered_map<std::wstring_view, uint32_t> features);

```

Comment:
Can we just use `UpdateFont` as the single point of entry of font settings? In my understanding the features are basically like weight & style & etc, right? I'm OK with `DidUserSetFeatures`, because it's read-only. But for anything that can change the internal properties in `DxFontRenderData`, I'd like to do it through `UpdateFont`.

So the idea is roughly:  move `_features` into `FontInfoDesired` &. Update `_desiredFont->features` in `ControlCore`. Pass the `features` through `UpdateFont`. Then continue.

This would be a less invasive way. BTW this also eliminates `DxEngine::UpdateFontFeatures`.

Reply:
Is `features` something we need in both `FontInfoBase` & `DxFontInfo`? If it is, then it calls for the merging of these classes. 

CC @miniksa for suggestions.

Reply:
We cannot move `_features` into `FontInfoDesired` or `FontInfo` because that part of the code is shared with conhost. The features need to be specific to only the DxEngine (similar to the way we do pixel shader path)

Reply:
No i mean technically all you need is a std::unordered_map, right? No DX structure related whatsoever. That is the low level representation. And you can convert that into DX structure in upper layer.

Unless there’s other regulations, then I am OK

Reply:
Sorry it took a while to get back to you on this!

Even though its a low level representation, conhost will never use the map in font info, so it is very wasteful to always ask conhost to allocate space for an empty map especially considering that the majority use case is conhost

Reply:
If it’s not used in conhost, I assume the compiler will maybe try to optimize them away, right? Even without optimization, I would personally trade the few bytes for architectural benefits without doubt. 

I am gonna leave the choice to you guys, especially @miniksa since he created all the DX classes.

Reply:
Nah. The compiler cannot optimize out members from a class. Memory layouts aren't malleable like that!

Reply:
We **could** do it as a `std::optional<std::unordered_map<std::wstring_view, uint32_t>>` right? I don't think that will alloc unless it is actually used?

Reply:
As discussed, we use `UpdateFont` now for the features and axes. `_SetFeatures` and `_SetAxes` are now just private helper functions inside `DxFontRenderData`